scripts/cnode-helper-scripts/cncli.sh

```diff
@@ -797,10 +807,203 @@ cncliPTsendslots() {
+    if ! getConsensus "${EPOCH}"; then echo "ERROR: Failed to fetch protocol parameters for epoch ${EPOCH}.; return 1"; fi
+     if [[ "$EPOCH" == "$curr_epoch" ]]; then
+       runCurrentEpoch
+     elif [[ "$EPOCH" == "$next_epoch" ]]; then
+       runNextEpoch
+     else
+       runPreviousEpochs
+     fi
+  done
+
+  id=1
+  while IFS= read -r row; do
+    echo "$id,$row" >> "$csvfile"
+    ((id++))
+  done < "$tmpcsv"
+
+  sqlite3 "$BLOCKLOG_DB" <<EOF
+DELETE FROM epochdata;
+VACUUM;
+.mode csv
+.import '$csvfile' epochdata
+REINDEX epochdata;
+EOF
+
+  row_count=$(sqlite3 "$BLOCKLOG_DB" "SELECT COUNT(*) FROM epochdata;")
+  echo "$row_count rows have been loaded into epochdata table in blocklog db"
+  echo "~ CNCLI epochdata table load completed ~"
+
+  rm $csvfile $tmpcsv
+}
+
+processSingleEpoch() {
+ if isNumber "$1"; then
+     getCurrNextEpoch
+     IFS=' ' read -r -a epochs_array <<< "$EPOCHS"
+
+   if [[ ! " ${epochs_array[@]} " =~ "$1" ]]; then
+     echo "No slots found in blocklog table for epoch ${EPOCH}."
+     echo
+     exit 1
+   fi
+
+   if ! getConsensus "${1}"; then echo "ERROR: Failed to fetch protocol parameters for epoch ${1}.; return 1"; fi
+
+   if [[ "$1" == "$curr_epoch" ]]; then
+      runCurrentEpoch
+   elif [[ "$1" == "$next_epoch" ]]; then
+      runNextEpoch
+   else
+      runPreviousEpochs
+   fi
+
+   ID=$(sqlite3 "$BLOCKLOG_DB" "SELECT max(id) + 1 FROM epochdata;")
+   csv_row=$(cat "$tmpcsv")
+   modified_csv_row="${ID},${csv_row}"
+   echo "$modified_csv_row" > "$onerow_csv"
+
+   sqlite3 "$BLOCKLOG_DB" "DELETE FROM epochdata WHERE epoch = '$1';"
+   sqlite3 "$BLOCKLOG_DB" <<EOF
+.mode csv
+.import "$onerow_csv" epochdata
+REINDEX epochdata;
+EOF
+   row_count=$(sqlite3 "$BLOCKLOG_DB" "SELECT COUNT(*) FROM epochdata WHERE epoch = '$1';")
+   echo "$row_count row has been loaded into epochdata table in blocklog db for epoch '$1'"
+   echo "~ CNCLI epochdata table load completed ~"
+   echo
+
+   rm $onerow_csv $tmpcsv
+ else
+   echo "No data found in $tmpcsv."
+   exit 1
+ fi
+}
+
+cncliEpochData() {
+  cncliInit
+  cncliParams="--db ${CNCLI_DB} --byron-genesis ${BYRON_GENESIS_JSON} --shelley-genesis ${GENESIS_JSON} --pool-id ${POOL_ID} --pool-vrf-skey ${POOL_VRF_SKEY} --tz UTC"
+  EPOCHS=$(sqlite3 "$BLOCKLOG_DB" "SELECT replace(group_concat(DISTINCT epoch ORDER BY epoch ASC), ',', ' ') AS EPOCHS FROM blocklog;")
+  csvdir=/tmp
+  tmpcsv="${csvdir}/epochdata_tmp.csv"
+  csvfile="${csvdir}/epochdata.csv"
+  onerow_csv="${csvdir}/one_epochdata.csv"
+  > "$tmpcsv"
+  > "$csvfile"
+  > "$onerow_csv"
+
+  proc_msg="~ CNCLI epochdata table load started ~"
+
+  getNodeMetrics
+
+  if ! cncliDBinSync; then
+    echo ${proc_msg}
+    echo "CNCLI DB out of sync :( [$(printf "%2.4f %%" ${cncli_sync_prog})] ... check cncli sync service!"
+    exit 1
+  else
+    echo ${proc_msg}
+    local epochSlot=${slot_in_epoch}
+    local waitFromEpochstart=21600 # if within 6 hours of new epoch don't run epochdata
```

Comment:
Need to make this network agnostic and not mainnet specific based on genesis params, ie percentage of slots in epoch or something.

Reply:
Hello @Scitz0 

you mentioned this on the first pr but i forgot it, I changed as

>    read_genesis
    local epochSlot=${slot_in_epoch}
    local epochLength=${EPOCH_LENGTH}
    local remainingSlots=$(( epochLength - epochSlot))
    local waitFromEpochstart=21600 # if within 6 hours of new epoch don't run epochdata

   > if [[ $(( epochLength - remainingSlots)) -le $waitFromEpochstart ]]; then

Reply:
It's waitFromEpochstart that's the problem. Guild network for example has 1h epochs, ie 3600 slots in total per epoch. That is what I meant by this needing to be dynamic and a percentage of total slots and not a static value of 6h.

Reply:
What is the reason for the 6h wait after epoch start by the way?

Reply:
A note around this, I would definitely not suggest adding any hard-coded entries shen it comes to slots.
The reason when it currently requires waiting from Koios endpoints pov when transitioning across into new epoch is that dbsync 13.2 only completed populating epoch stake for current epoch few hours into new one (wasnt as high as 6 hours, but was certainly a bit).
With dbsync 13.3 (that will be used as part of Koios 1.2.0), this data is populated about 50-70% prior to completion of epoch instead of post, thus - this should be a non-issue post the upgrade.
Given that current behaviour is not changing, I'd prefer not to add such limits for now, we'd be trying to push for koios 1.2.0 sooner (in coming weeks)